File: java/src/main/java/toolgood/algorithm/internals/functions/mathsum2/Function_ERF.java
package toolgood.algorithm.internals.functions.mathsum2;

import java.math.BigDecimal;
import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_1;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.system.MathEx;

public final class Function_ERF extends Function_1 {
    public Function_ERF(FunctionBase func1) {
        super(func1);
    }

    public Function_ERF(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Erf";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        BigDecimal x = args1.NumberValue();
        return Operand.Create(Erf(x));
    }

    private static BigDecimal Erf(BigDecimal x) {
        final BigDecimal a1 = new BigDecimal("0.254829592");
        final BigDecimal a2 = new BigDecimal("-0.284496736");
        final BigDecimal a3 = new BigDecimal("1.421413741");
        final BigDecimal a4 = new BigDecimal("-1.453152027");
        final BigDecimal a5 = new BigDecimal("1.061405429");
        final BigDecimal p = new BigDecimal("0.3275911");

        int sign = x.compareTo(BigDecimal.ZERO) < 0 ? -1 : 1;
        x = x.abs();

        if (x.compareTo(new BigDecimal("6.0")) > 0) {
            return new BigDecimal(sign);
        }

        BigDecimal t = BigDecimal.ONE.divide(BigDecimal.ONE.add(p.multiply(x)), java.math.MathContext.DECIMAL128);
        BigDecimal y = BigDecimal.ONE.subtract((((((a5.multiply(t).add(a4)).multiply(t)).add(a3)).multiply(t).add(a2)).multiply(t).add(a1)).multiply(t).multiply(MathEx.Exp(x.negate().multiply(x))));

        return new BigDecimal(sign).multiply(y);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}

File: JavaScript/src/Internals/Functions/MathBase/Function_ERF.js
import { Function_1 } from '../Function_1.js';
import { Operand } from '../../../Operand.js';

class Function_ERF extends Function_1 {
    get Name() {
        return 'Erf';
    }

    constructor(a) {
        super(a);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) { return args1; }
        let x = args1.NumberValue;
        return Operand.Create(this.erf(x));
    }

    erf(x) {
        // 用高精度补余误差函数合成,避免 A&S 逼近在大 x 时精度丢失
        return x >= 0 ? 1.0 - this.erfc(x) : this.erfc(-x) - 1.0;
    }

    erfc(x) {
        // Numerical Recipes 6.2 erfcc:对 |x| 做有理逼近
        let z = Math.abs(x);
        let t = 1.0 / (1.0 + 0.5 * z);
        let ans = t * Math.exp(-z * z - 1.26551223 + t * (1.00002368 + t * (0.37409196 + t * (0.09678418 + t * (-0.18628806 + t * (0.27886807 + t * (-1.13520398 + t * (1.48851587 + t * (-0.82215223 + t * 0.17087277)))))))));
        return x >= 0 ? ans : 2.0 - ans;
    }
}

export { Function_ERF };
